src/modulefinder/sentry_modulefinder_apple.c

```diff
@@ -130,19 +130,39 @@ remove_image(const struct mach_header *mh, intptr_t UNUSED(vmaddr_slide))
 sentry_value_t
 sentry_get_modules_list(void)
 {
+    // We have 2 locked blocks here (actually 3, with the one inside of the
+    // `add_image` callback). We do that because we have observed deadlocks when
+    // code concurrently `dlopen`s and thus invokes the `add_image` callback
+    // from a different thread.
+    bool should_initialize = false;
     sentry__mutex_lock(&g_mutex);
-    if (!g_initialized) {
+    if (g_initialized) {
+        // avoid an unlock/lock cycle in this case
+        goto is_initialized;
+    } else {
         g_modules = sentry_value_new_list();
+        g_initialized = true;
+        should_initialize = true;
+    }
+    sentry__mutex_unlock(&g_mutex);
+
+    if (should_initialize) {
         // TODO: maybe use `_dyld_image_count` and `_dyld_get_image_header`?
-        // Those functions are documented to not be thread-safe, though using
-        // the `register_X` functions are also unsafe because they lack a
-        // corresponding `unregister` function, and will thus crash when sentry
-        // itself is unloaded.
+        // Those functions are documented to not be thread-safe, though
+        // using the `register_X` functions are also unsafe because they
+        // lack a corresponding `unregister` function, and will thus crash
+        // when sentry itself is unloaded.
         _dyld_register_func_for_add_image(add_image);
         _dyld_register_func_for_remove_image(remove_image);
-        g_initialized = true;
     }
-    sentry_value_t modules = g_modules;
+
+    sentry__mutex_lock(&g_mutex);
+    // now that is strange, having this declaration after the goto label gives a
+    // `expected expression` error.
+    sentry_value_t modules;
```

Comment:
You should move that even before the `goto` statement, rather than just the label. Just put this at the very top of the function.